src/utils.js

```diff
@@ -1,25 +1,50 @@
+import path from "path-browserify";
 import parse from "url-parse";
 
 export const defaultPortalUrl = "https://siasky.net";
 
-export function open(portalUrl, skylink) {
-  const url = getUrl(portalUrl, parseSkylink(skylink));
+export const options = {
+  portalEndpointPath: "",
+  // TODO:
+  // customUserAgent: "",
+};
 
-  window.open(url, "_blank");
+function getFilePath(file) {
+  return file.webkitRelativePath || file.path || file.name;
 }
 
-export function getUrl(portalUrl, skylink, options = {}) {
-  const parsed = parse(portalUrl);
+export function getRelativeFilePath(file) {
+  const filePath = getFilePath(file);
+  const { root, dir, base } = path.parse(filePath);
+  const relative = path.normalize(dir).slice(root.length).split(path.sep).slice(1);
+
+  return path.join(...relative, base);
+}
 
-  parsed.set("pathname", parseSkylink(skylink));
+export function getRootDirectory(file) {
+  const filePath = getFilePath(file);
+  const { root, dir } = path.parse(filePath);
 
-  if (options.download) {
-    parsed.set("query", { attachment: true });
-  }
+  return path.normalize(dir).slice(root.length).split(path.sep)[0];
+}
 
+export function makeUrl(portalUrl, endpointPath, query = {}) {
+  const parsed = parse(portalUrl);
+  parsed.set("pathname", endpointPath);
+  parsed.set("query", query);
   return parsed.toString();
 }
 
+export function makeUrlWithSkylink(portalUrl, endpointPath, skylink, query = {}) {
+  skylink = parseSkylink(skylink);
+  // Right-trim any forward slashes from endpoint path.
+  endpointPath = endpointPath.replace(/\/+$/g, "");
+  // Left-trim any forward slashes from skylink.
+  skylink = skylink.replace(/^\/+/g, "");
+  endpointPath = `${endpointPath}/${skylink}`;
+  return makeUrl(portalUrl, endpointPath, query);
```

Comment:
I think it could be just this because `path.poisx.join` will take care of any excessive `/` between `endpointPath` and `skylink` and `makeUrl` will properly merge the pathname with the hostname.
```suggestion
  const parsedSkylink = parseSkylink(skylink);
  
  return makeUrl(portalUrl, path.posix.join(endpointPath, parsedSkylink), query);
```